ui/vr_manager.js

```diff
@@ -0,0 +1,496 @@
+      if (typeof DeviceMotionEvent.requestPermission == 'function') {
+        const userGestureListener = () => {
+          DeviceMotionEvent.requestPermission().then((newPermissionState) => {
+            if (newPermissionState !== 'granted' ||
+                deviceOrientationListener) {
+              return;
+            }
+            deviceOrientationListener = true;
+            this.setupDeviceOrientationListener_();
+          });
+        };
+        DeviceMotionEvent.requestPermission().then((permissionState) => {
+          this.eventManager_.unlisten(
+              this.container_, 'click', userGestureListener);
+          this.eventManager_.unlisten(
+              this.container_, 'mouseup', userGestureListener);
+          if (navigator.maxTouchPoints > 0) {
+            this.eventManager_.unlisten(
+                this.container_, 'touchend', userGestureListener);
+          }
+          if (permissionState !== 'granted') {
+            this.eventManager_.listenOnce(
+                this.container_, 'click', userGestureListener);
+            this.eventManager_.listenOnce(
+                this.container_, 'mouseup', userGestureListener);
+            if (navigator.maxTouchPoints > 0) {
+              this.eventManager_.listenOnce(
+                  this.container_, 'touchend', userGestureListener);
+            }
+            return;
+          }
+          deviceOrientationListener = true;
+          this.setupDeviceOrientationListener_();
+        }).catch(() => {
+          this.eventManager_.unlisten(
+              this.container_, 'click', userGestureListener);
+          this.eventManager_.unlisten(
+              this.container_, 'mouseup', userGestureListener);
+          if (navigator.maxTouchPoints > 0) {
+            this.eventManager_.unlisten(
+                this.container_, 'touchend', userGestureListener);
+          }
+          this.eventManager_.listenOnce(
+              this.container_, 'click', userGestureListener);
+          this.eventManager_.listenOnce(
+              this.container_, 'mouseup', userGestureListener);
+          if (navigator.maxTouchPoints > 0) {
+            this.eventManager_.listenOnce(
+                this.container_, 'touchend', userGestureListener);
+          }
+        });
+      } else {
+        deviceOrientationListener = true;
+        this.setupDeviceOrientationListener_();
+      }
+    }
+  }
+
+  /**
+   * @private
+   */
+  setupDeviceOrientationListener_() {
+    this.eventManager_.listen(window, 'deviceorientation', (e) => {
+      const event = /** @type {!DeviceOrientationEvent} */(e);
+      let alphaDif = (event.alpha || 0) - this.prevAlpha_;
+      let betaDif = (event.beta || 0) - this.prevBeta_;
+      let gammaDif = (event.gamma || 0) - this.prevGamma_;
+
+      if (Math.abs(alphaDif) > 10) {
+        alphaDif = 0;
+      }
+      if (Math.abs(gammaDif) > 10) {
```

Comment:
Also, is 10 the correct value for all three of these?
Alpha and beta both have 360 degrees in total, but gamma only goes from -90 to 90 (180 degrees in total). I would think that would mean that a large gamma movement would be 5 degrees, not 10?

Reply:
Yes, it was a mistake, I'm going to change it.